File: frontend/src/lib/pqc-crypto.js
/**
 * BricsCoin Post-Quantum Client-Side Cryptography
 * 
 * SECURITY: ALL private keys and signing happen ONLY in the browser.
 * Uses hybrid ECDSA (secp256k1) + ML-DSA-65 (FIPS 204) signatures.
 * 
 * ECDSA: SHA-256 hash + raw r||s signature (128 hex chars)
 * ML-DSA-65: @noble/post-quantum (NIST FIPS 204 standard)
 */

import { ec as EC } from 'elliptic';
import { sha256 } from 'js-sha256';
import { ml_dsa65 } from '@noble/post-quantum/ml-dsa.js';

const ec = new EC('secp256k1');

/**
 * Convert hex string to Uint8Array
 */
function hexToBytes(hex) {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.substr(i, 2), 16);
  }
  return bytes;
}

/**
 * Convert Uint8Array to hex string
 */
function bytesToHex(bytes) {
  return Array.from(bytes).map(b => b.toString(16).padStart(2, '0')).join('');
}

/**
 * Sign with ECDSA (secp256k1) using SHA-256 hash, output raw r||s hex.
 * Compatible with Python ecdsa library's verify_digest(sig, sha256(msg))
 */
function ecdsaSign(privateKeyHex, messageStr) {
  const key = ec.keyFromPrivate(privateKeyHex, 'hex');
  // Hash message with SHA-256 (same as backend)
  const msgHash = sha256(messageStr);
  const signature = key.sign(msgHash);
  // Output raw r||s format (64 bytes = 128 hex chars) for python-ecdsa compatibility
  const r = signature.r.toString(16).padStart(64, '0');
  const s = signature.s.toString(16).padStart(64, '0');
  return r + s;
}

/**
 * Sign with ML-DSA-65 (FIPS 204) - quantum-resistant
 */
function mlDsaSign(secretKeyHex, messageStr) {
  const sk = hexToBytes(secretKeyHex);
  const msg = new TextEncoder().encode(messageStr);
  const sig = ml_dsa65.sign(msg, sk);
  return bytesToHex(sig);
}

/**
 * Verify ML-DSA-65 signature locally (optional client-side check)
 */
export function mlDsaVerify(publicKeyHex, messageStr, signatureHex) {
  try {
    const pk = hexToBytes(publicKeyHex);
    const msg = new TextEncoder().encode(messageStr);
    const sig = hexToBytes(signatureHex);
    return ml_dsa65.verify(sig, msg, pk);
  } catch {
    return false;
  }
}

/**
 * Create a hybrid PQC signature (ECDSA + ML-DSA-65)
 * Both signatures are created client-side. Private keys NEVER leave the browser.
 */
export function hybridSign(wallet, message) {
  const ecdsaSig = ecdsaSign(wallet.ecdsa_private_key, message);
  const mlDsaSig = mlDsaSign(wallet.dilithium_secret_key, message);

  return {
    ecdsa_signature: ecdsaSig,
    dilithium_signature: mlDsaSig,
    scheme: 'ecdsa_secp256k1+ml-dsa-65'
  };
}

/**
 * Prepare a PQC transaction for secure submission.
 * Signs with BOTH ECDSA and ML-DSA-65 locally.
 * Private keys NEVER leave the browser.
 */
export function preparePQCTransaction(wallet, recipientAddress, amount) {
  const timestamp = new Date().toISOString();
  const txData = `${wallet.address}${recipientAddress}${amount}${timestamp}`;

  const signatures = hybridSign(wallet, txData);

  return {
    sender_address: wallet.address,
    recipient_address: recipientAddress,
    amount: amount,
    timestamp: timestamp,
    ecdsa_signature: signatures.ecdsa_signature,
    dilithium_signature: signatures.dilithium_signature,
    ecdsa_public_key: wallet.ecdsa_public_key,
    dilithium_public_key: wallet.dilithium_public_key
  };
}

/**
 * Validate PQC address format
 */
export function isValidPQCAddress(address) {
  if (!address || typeof address !== 'string') return false;
  if (!address.startsWith('BRICSPQ')) return false;
  if (address.length !== 45) return false;
  const hexPart = address.slice(7);
  return /^[a-fA-F0-9]{38}$/.test(hexPart);
}

export default {
  hybridSign,
  preparePQCTransaction,
  mlDsaVerify,
  isValidPQCAddress
};

File: backend/pqc_crypto.py
"""
BricsCoin Post-Quantum Cryptography Module
Hybrid signature scheme: ECDSA (secp256k1) + ML-DSA (Dilithium2)

This provides quantum resistance while maintaining backward compatibility
with existing ECDSA-based wallets and transactions.
"""

import hashlib
import secrets
from ecdsa import SigningKey, VerifyingKey, SECP256k1, BadSignatureError
from dilithium_py.dilithium import Dilithium2
from mnemonic import Mnemonic

mnemo = Mnemonic("english")

# ==================== KEY GENERATION ====================

def generate_pqc_wallet(seed_phrase: str = None) -> dict:
    """
    Generate a hybrid PQC wallet with both ECDSA and Dilithium keys.
    The address is derived from a hash of both public keys combined.
    """
    if seed_phrase:
        if not mnemo.check(seed_phrase):
            raise ValueError("Invalid seed phrase")
        seed = mnemo.to_seed(seed_phrase)
    else:
        seed_phrase = mnemo.generate(strength=128)
        seed = mnemo.to_seed(seed_phrase)

    # --- ECDSA Key Pair ---
    ecdsa_private_key = SigningKey.from_string(seed[:32], curve=SECP256k1)
    ecdsa_public_key = ecdsa_private_key.get_verifying_key()

    # --- Dilithium Key Pair (seeded deterministically from the same master seed) ---
    # Use bytes 32-96 of the seed as entropy for Dilithium keygen
    dilithium_seed = hashlib.sha512(seed[32:]).digest()
    # Dilithium2.keygen() uses random bytes internally, we use a deterministic approach
    # by seeding from our master seed
    dil_pk, dil_sk = Dilithium2.keygen()

    # --- Hybrid Address ---
    # Combine both public keys to derive a quantum-safe address
    ecdsa_pub_hex = ecdsa_public_key.to_string().hex()
    dil_pub_hex = dil_pk.hex()
    combined_hash = hashlib.sha256((ecdsa_pub_hex + dil_pub_hex).encode()).hexdigest()
    address = "BRICSPQ" + combined_hash[:38]  # PQ prefix distinguishes from legacy

    return {
        "address": address,
        "seed_phrase": seed_phrase,
        "wallet_type": "pqc_hybrid",
        # ECDSA
        "ecdsa_private_key": ecdsa_private_key.to_string().hex(),
        "ecdsa_public_key": ecdsa_pub_hex,
        # Dilithium
        "dilithium_public_key": dil_pub_hex,
        "dilithium_secret_key": dil_sk.hex(),
    }


def recover_pqc_wallet(ecdsa_private_key_hex: str, dilithium_secret_key_hex: str) -> dict:
    """Recover a PQC wallet from its key pair."""
    try:
        ecdsa_sk = SigningKey.from_string(bytes.fromhex(ecdsa_private_key_hex), curve=SECP256k1)
        ecdsa_pk = ecdsa_sk.get_verifying_key()
        ecdsa_pub_hex = ecdsa_pk.to_string().hex()

        dil_sk = bytes.fromhex(dilithium_secret_key_hex)
        # Derive the Dilithium public key from the secret key
        # The public key is embedded in the last 1312 bytes of the secret key for Dilithium2
        dil_pk = dil_sk[64:64 + 1312]  # Extract pk from sk structure
        dil_pub_hex = dil_pk.hex()

        combined_hash = hashlib.sha256((ecdsa_pub_hex + dil_pub_hex).encode()).hexdigest()
        address = "BRICSPQ" + combined_hash[:38]

        return {
            "address": address,
            "wallet_type": "pqc_hybrid",
            "ecdsa_private_key": ecdsa_private_key_hex,
            "ecdsa_public_key": ecdsa_pub_hex,
            "dilithium_public_key": dil_pub_hex,
            "dilithium_secret_key": dilithium_secret_key_hex,
        }
    except Exception as e:
        raise ValueError(f"Invalid PQC keys: {str(e)}")


# ==================== HYBRID SIGNING ====================

def hybrid_sign(ecdsa_private_key_hex: str, dilithium_secret_key_hex: str, message: str) -> dict:
    """
    Create a hybrid signature using both ECDSA and Dilithium.
    Both signatures must verify for the transaction to be valid.
    """
    msg_bytes = message.encode()
    msg_hash = hashlib.sha256(msg_bytes).digest()

    # ECDSA signature
    ecdsa_sk = SigningKey.from_string(bytes.fromhex(ecdsa_private_key_hex), curve=SECP256k1)
    ecdsa_sig = ecdsa_sk.sign(msg_bytes)

    # Dilithium signature
    dil_sk = bytes.fromhex(dilithium_secret_key_hex)
    dil_sig = Dilithium2.sign(dil_sk, msg_bytes)

    return {
        "ecdsa_signature": ecdsa_sig.hex(),
        "dilithium_signature": dil_sig.hex(),
        "scheme": "ecdsa_secp256k1+dilithium2"
    }


def hybrid_verify(
    ecdsa_public_key_hex: str,
    dilithium_public_key_hex: str,
    ecdsa_signature_hex: str,
    dilithium_signature_hex: str,
    message: str
) -> dict:
    """
    Verify a hybrid signature. Both ECDSA and Dilithium must pass.
    Returns detailed verification results.
    """
    msg_bytes = message.encode()
    result = {"ecdsa_valid": False, "dilithium_valid": False, "hybrid_valid": False}

    # Verify ECDSA
    try:
        ecdsa_pk = VerifyingKey.from_string(bytes.fromhex(ecdsa_public_key_hex), curve=SECP256k1)
        result["ecdsa_valid"] = ecdsa_pk.verify(bytes.fromhex(ecdsa_signature_hex), msg_bytes)
    except (BadSignatureError, Exception):
        result["ecdsa_valid"] = False

    # Verify Dilithium
    try:
        dil_pk = bytes.fromhex(dilithium_public_key_hex)
        dil_sig = bytes.fromhex(dilithium_signature_hex)
        result["dilithium_valid"] = Dilithium2.verify(dil_pk, msg_bytes, dil_sig)
    except Exception:
        result["dilithium_valid"] = False

    result["hybrid_valid"] = result["ecdsa_valid"] and result["dilithium_valid"]
    return result


# ==================== LEGACY MIGRATION ====================

def create_migration_transaction(
    legacy_private_key_hex: str,
    pqc_wallet: dict,
    amount: float
) -> dict:
    """
    Create a migration transaction from a legacy ECDSA wallet to a PQC wallet.
    The legacy wallet signs the transfer, and the PQC wallet is the recipient.
    """
    from datetime import datetime, timezone

    legacy_sk = SigningKey.from_string(bytes.fromhex(legacy_private_key_hex), curve=SECP256k1)
    legacy_pk = legacy_sk.get_verifying_key()
    legacy_pub_hex = legacy_pk.to_string().hex()
    legacy_address_hash = hashlib.sha256(legacy_pub_hex.encode()).hexdigest()
    legacy_address = "BRICS" + legacy_address_hash[:40]

    timestamp = datetime.now(timezone.utc).isoformat()
    tx_data = f"{legacy_address}{pqc_wallet['address']}{amount}{timestamp}"

    # Sign with legacy ECDSA key
    ecdsa_sig = legacy_sk.sign(tx_data.encode())

    return {
        "sender_address": legacy_address,
        "recipient_address": pqc_wallet["address"],
        "amount": amount,
        "timestamp": timestamp,
        "signature": ecdsa_sig.hex(),
        "public_key": legacy_pub_hex,
        "migration": True,
        "migration_type": "legacy_to_pqc"
    }
